### ISSUE TITLE
Remove `hashbrown` dependency

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#1444]
I was curious if it's possible to make `hashbrown` optional and I concluded it's likely possible to remove it entirely.

It is only used in BIP-158 implementation where it implements de-duplication of items. Those items are then hashed (potentially creating collisions), sorted and encoded to a compressed binary format. From what I understood the operations on the filter never take into account the number of occurrences of the item in the filter. The hashes of queries are even de-duplicated and only consider presence of the element in the set, not count.

From this I conclude that de-duplicating elements rather than hashes is a bug. And if we're going to sort them anyway we may as well put them inside `BTreeSet` rather than `HashSet`. `BTreeSet` is in `alloc` so no need for `hashbrown`. I guess the whole thing should also become faster after this change since we also eliminate allocations.

Also curiously there are two approaches to de-duplication: use of `HashSet` and `Vec` with `sort_unstable` + `deduplicate`. My suspicion is that `Vec` is faster but uses more memory but not sure. Should we use the same approach in all places or is there some (undocumented) good reason to use different approaches?

**Comment by tcharding** [comment]
#1576 removes the `hashbrown` dependency but I did not investigate the possible bugs described in this issue.

**Comment by Kixunil** [comment]
Cc @devrandom you created the BIP158 module, could you please comment on this issue?

**Comment by devrandom** [comment]
> Cc @devrandom you created the BIP158 module, could you please comment on this issue?

I did not create it, Tamas Blummer did.  But the analysis above sounds correct.

**Comment by tcharding** [comment]
Ok, with those comments I'll work a little hard - thanks for nudging me.

**Comment by tcharding** [comment]
Hey @Kixunil I took a look at `bip158` and I disagree a bit with your analysis. Firstly I think you are correct in that AFAICT the reference golang version [0] does not remove duplicates so the encoded filter will be different from the one we produce. I think this is a bug in the reference implementation not ours. It seems to me our implementation will produce a smaller encoded filter that answers queries correctly, so that's a win. We should probably test that as well as implementing the test vectors from the bip.

I asked a question on the golang version.

[0] https://github.com/Roasbeef/btcutil/tree/gcs/gcs

**Comment by tcharding** [comment]
After comments https://github.com/btcsuite/btcd/issues/1940 I believe our code is correct and this issue can be closed. See https://github.com/btcsuite/btcd/blob/016b3adb3688651d13af7cbafb514d8947e68440/btcutil/gcs/builder/builder.go#L153-L163 for the golang equivalent.

@Kixunil please close if you are satisfied.

**Comment by tcharding** [comment]
And FTR I think this Bitcoin Core code confirms the same thing (`blockfilter.cpp`)
```c++
static GCSFilter::ElementSet BasicFilterElements(const CBlock& block,
                                                 const CBlockUndo& block_undo)
{
    GCSFilter::ElementSet elements;

    for (const CTransactionRef& tx : block.vtx) {
        for (const CTxOut& txout : tx->vout) {
            const CScript& script = txout.scriptPubKey;
            if (script.empty() || script[0] == OP_RETURN) continue;
            elements.emplace(script.begin(), script.end());
        }
    }

    for (const CTxUndo& tx_undo : block_undo.vtxundo) {
        for (const Coin& prevout : tx_undo.vprevout) {
            const CScript& script = prevout.out.scriptPubKey;
            if (script.empty()) continue;
            elements.emplace(script.begin(), script.end());
        }
    }

    return elements;
}
```

**Comment by Kixunil** [comment]
Yeah, this is done, but someone should take a look at those collisions. I don't have much time right now.

**Comment by tcharding** [comment]
From my understanding false positives are a part of the compact block filters anyways, I don't see why collisions matter for data included in the filter.

**Comment by Kixunil** [comment]
We filter-out equal elements, not equal hashes, so collisions will end up in the filter as zeros.

**Comment by tcharding** [comment]
Oh, sorry for being slow, I get you now, I'll go back to the Bitcoin Core code and the golang code and see what they do. Unless you already did this?

**Comment by Kixunil** [comment]
I didn't, doing some other crap lately.

**Comment by tcharding** [comment]
No stress, I'll check it out.

**Comment by tcharding** [comment]
Looks to me like Bitcoin Core does it the same way we do, ignoring hash colisions.
```
std::vector<uint64_t> GCSFilter::BuildHashedSet(const ElementSet& elements) const
{
    std::vector<uint64_t> hashed_elements;
    hashed_elements.reserve(elements.size());
    for (const Element& element : elements) {
        hashed_elements.push_back(HashToRange(element));
    }
    std::sort(hashed_elements.begin(), hashed_elements.end());
    return hashed_elements;
}
```
and ```
    uint64_t last_value = 0;
    for (uint64_t value : BuildHashedSet(elements)) {
        uint64_t delta = value - last_value;
        GolombRiceEncode(bitwriter, m_params.m_P, delta);
        last_value = value;
    }
```

**Comment by Kixunil** [comment]
Yeah, opened an issue there.
